### PR TITLE
Improve environment variable parameter parsing

### LINKED ISSUE #445: Unable to parse variable string
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I'm trying to pass some weird env variables for a Laravel app but I get "Unable to parse variable string".

```
nixpacks build . --name testing --env "APP_KEY=base64:abcdef...="
```

It doesn't like the final `=` character.

### To reproduce

_No response_

### Expected behavior

_No response_

### Environment

-

### PR DESCRIPTION
**Pull request by Milo123459**
## What does this PR address?

Fixes #445

Improves parsing by using a Regex.

### CODE REVIEW

**Comment by nebulatgs** [comment]
be more descriptive in pull requests (at least summarize what its fixing)

File: src/nixpacks/environment.rs [hunk 1]
```diff
@@ -16,19 +17,26 @@ impl Environment {
     pub fn from_envs(envs: Vec<&str>) -> Result<Environment> {
         let mut environment = Environment::default();
         for env in envs {
-            let v: Vec<&str> = env.split('=').collect();
-            if v.len() == 1 {
+            let matches = Regex::new(r"([^=]*)(?:=?)([^=]*)([\s\S]*)")
+                .unwrap()
+                .captures(env)
+                .unwrap();
```
Comment (coffee-cup) [inline src/nixpacks/environment.rs:23]: Should not `unwrap` here. Or anywhere below either.

File: src/nixpacks/environment.rs [hunk 2]
```diff
@@ -16,19 +17,26 @@ impl Environment {
     pub fn from_envs(envs: Vec<&str>) -> Result<Environment> {
         let mut environment = Environment::default();
         for env in envs {
-            let v: Vec<&str> = env.split('=').collect();
-            if v.len() == 1 {
+            let matches = Regex::new(r"([^=]*)(?:=?)([^=]*)([\s\S]*)")
```
Comment (coffee-cup) [inline src/nixpacks/environment.rs:20]: I don't think this regex is correct. Environment variable names cannot have any character. We should limit it to alphanum characters

**Review by coffee-cup — COMMENTED** [review]
Please add some tests for the new variable parsing

File: src/nixpacks/environment.rs (hunk 2 above)
Comment (Milo123459) [reply src/nixpacks/environment.rs:20]: I was mirroring behaviour that was already there, which was splitting which would account for any character. But I'll implement.

File: src/nixpacks/environment.rs (hunk 1 above)
Comment (Milo123459) [reply src/nixpacks/environment.rs:23]: In my tests, there can never be less then 1 match so it's fine to unwrap.

File: src/nixpacks/environment.rs [hunk 3]
```diff
@@ -16,19 +17,26 @@ impl Environment {
     pub fn from_envs(envs: Vec<&str>) -> Result<Environment> {
         let mut environment = Environment::default();
         for env in envs {
-            let v: Vec<&str> = env.split('=').collect();
-            if v.len() == 1 {
+            let matches = Regex::new(r"([A-Za-z0-9_-]*)(?:=?)([A-Za-z0-9_-]*)([\s\S]*)")
```
Comment (nebulatgs) [inline src/nixpacks/environment.rs:20]:
```suggestion
            let matches = Regex::new(r"([A-Za-z0-9_-]*)(?:=?)([^=]*)([\s\S]*)")
```
the variable body can be anything that isnt an `=`

**Review by nebulatgs — DISMISSED** [review]

File: src/nixpacks/environment.rs [hunk 4]
```diff
@@ -86,11 +90,6 @@ mod tests {
         assert!(environment.get_variable("NON_EXISTANT").is_none());
     }
 
-    #[test]
-    fn test_create_invalid_environment() {
-        assert!(Environment::from_envs(vec!["INVALID=ENV=CONFIG"]).is_err());
-    }
-
```
Comment (coffee-cup) [inline src/nixpacks/environment.rs:93]: Instead of removing this test can you instead assert that it is valid and that the name/value are correct

Comment (Milo123459) [reply src/nixpacks/environment.rs]: done

**Review by coffee-cup — APPROVED** [review]

File: src/nixpacks/environment.rs (hunk 3 above)
Comment (Milo123459) [reply src/nixpacks/environment.rs]: done

**Pull request merged**